cdap-examples/WikipediaPipeline/src/main/java/co/cask/cdap/examples/wikipedia/WikipediaPipelineWorkflow.java

```diff
@@ -0,0 +1,144 @@
+    addMapReduce(WikipediaPipelineApp.LIKES_TO_DATASET_MR_NAME);
+    condition(new EnoughDataToProceed())
+      .condition(new IsWikipediaSourceOnline())
+        .addAction(new DownloadWikiDataAction())
+      .otherwise()
+        .addMapReduce(WikipediaPipelineApp.WIKIPEDIA_TO_DATASET_MR_NAME)
+      .end()
+      .addMapReduce(WikiContentValidatorAndNormalizer.NAME)
+      .fork()
+        .addSpark(SparkWikipediaAnalyzer.NAME)
+      .also()
+        .addMapReduce(TopNMapReduce.NAME)
+      .join()
+    .otherwise()
+    .end();
+  }
+
+  static class EnoughDataToProceed implements Predicate<WorkflowContext> {
+
+    @Override
+    public boolean apply(WorkflowContext context) {
+      Map<String, String> runtimeArguments = context.getRuntimeArguments();
+      int threshold = 10;
+      if (runtimeArguments.containsKey("min.page.threshold")) {
+        threshold = Integer.parseInt(runtimeArguments.get("min.page.threshold"));
+      }
+      WorkflowToken token = context.getToken();
+      Value result = token.get("result", WikipediaPipelineApp.LIKES_TO_DATASET_MR_NAME);
+      Value numPages = token.get("custom.num.records", WorkflowToken.Scope.SYSTEM);
+      boolean conditionResult = result != null && result.getAsBoolean() &&
+        numPages != null && numPages.getAsLong() > threshold;
+      token.put("result", Value.of(conditionResult));
+
+      // Also add information in the token on whether to download wikipedia data over the internet
+      // NOTE: The following predicate can even consume this information through runtimeArguments. However,
+      // we want to demonstrate the usage of workflow token through this example, so adding this indirection.
+      if (runtimeArguments.containsKey("mode")) {
+        token.put("online", Value.of(runtimeArguments.get("mode").equalsIgnoreCase("online")));
+      }
+      return conditionResult;
+    }
+  }
+
+  static class IsWikipediaSourceOnline implements Predicate<WorkflowContext> {
+
+    @Override
+    public boolean apply(WorkflowContext context) {
+      WorkflowToken token = context.getToken();
+      Value online = token.get("online", EnoughDataToProceed.class.getSimpleName(), WorkflowToken.Scope.USER);
+      return online != null && online.getAsBoolean();
+    }
+  }
+
+  private static class DownloadWikiDataAction extends AbstractWorkflowAction {
+    private static final Logger LOG = LoggerFactory.getLogger(DownloadWikiDataAction.class);
+
+    @UseDataSet(WikipediaPipelineApp.PAGE_TITLES_DATASET)
+    private KeyValueTable pages;
+
+    @UseDataSet(WikipediaPipelineApp.RAW_WIKIPEDIA_DATASET)
+    private KeyValueTable wikiData;
+
+    @Override
+    public void run() {
+      CloseableIterator<KeyValue<byte[], byte[]>> scanner = pages.scan(null, null);
+      while (scanner.hasNext()) {
+        KeyValue<byte[], byte[]> next = scanner.next();
+        String rawWikiJson;
+        try {
+          rawWikiJson = downloadWikiData(Bytes.toString(next.getValue()));
```

Comment:
even if there are say 100k pages, if we try to request everything in a second, we could be blocked right?. rather it will be good to make the allowed number of requests per second and sleep and repeat. what do you say ?